py/src/braintrust/_types.py

```diff
@@ -352,4 +352,147 @@ class AttachmentStatus(TypedDict):
     """
 
 
+IfExists = Literal["error", "ignore", "replace"]
+
+
+class SavedFunctionId1(TypedDict):
+    type: Literal["function"]
+    id: str
+
+
+class SavedFunctionId2(TypedDict):
+    type: Literal["global"]
+    name: str
+
+
+SavedFunctionId = Union[SavedFunctionId1, SavedFunctionId2]
+
+
+class Function1(TypedDict):
+    name: str
+    description: NotRequired[Optional[str]]
+    parameters: NotRequired[Optional[Mapping[str, Any]]]
+    strict: NotRequired[Optional[bool]]
+
+
+class ToolFunctionDefinition(TypedDict):
+    type: Literal["function"]
+    function: Function1
+
+
+class ChatCompletionContentPartText(TypedDict):
+    text: NotRequired[Optional[str]]
+    type: Literal["text"]
+
+
+class ImageUrl(TypedDict):
+    url: str
+    detail: NotRequired[Optional[Union[Literal["auto"], Literal["low"], Literal["high"]]]]
+
+
+class ChatCompletionContentPartImage(TypedDict):
+    image_url: ImageUrl
+    type: Literal["image_url"]
+
+
+ChatCompletionContentPart = Union[ChatCompletionContentPartText, ChatCompletionContentPartImage]
+
+
+ChatCompletionContent = Union[str, Sequence[ChatCompletionContentPart]]
+
+
+class Function2(TypedDict):
+    arguments: str
+    name: str
+
+
+class ChatCompletionMessageToolCall(TypedDict):
+    id: str
+    function: Function2
+    type: Literal["function"]
+
+
+class ChatCompletionMessageParam1(TypedDict):
+    content: NotRequired[Optional[str]]
+    role: Literal["system"]
+    name: NotRequired[Optional[str]]
+
+
+class ChatCompletionMessageParam2(TypedDict):
+    content: NotRequired[Optional[ChatCompletionContent]]
+    role: Literal["user"]
+    name: NotRequired[Optional[str]]
+
+
+class FunctionCall1(TypedDict):
+    arguments: str
+    name: str
+
+
+class ChatCompletionMessageParam3(TypedDict):
```

Comment:
I take it this is to avoid coupling the sdk with a particular openai version? I noticed something similar in autoevals

Reply:
Yeah, and to provide the strong typing from the zod types in JS: you can see the generated type list https://github.com/braintrustdata/braintrust/pull/3776/files#diff-14c2b65c214476c421b4e1ecaaacd5ee481b5918ce3f9352db120f02be03ac3bR59